File: app/routes/news.py
from datetime import datetime
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required
from app.models import News
from app import db

news_bp = Blueprint("news", __name__)


@news_bp.route("", methods=["GET"])
def get_news():
    lang = request.args.get("lang", "en")
    page = request.args.get("page", 1, type=int)
    per_page = request.args.get("per_page", 10, type=int)
    player_id = request.args.get("player_id", type=int)

    query = News.query.filter_by(published=True)

    # Filter by region: show news for 'both' + the requested lang
    query = query.filter(db.or_(News.region == lang, News.region == "both"))

    if player_id:
        query = query.filter_by(player_id=player_id)

    query = query.order_by(db.case((News.published_at.is_(None), 1), else_=0), News.published_at.desc())
    pagination = query.paginate(page=page, per_page=per_page, error_out=False)

    return jsonify({
        "news": [n.to_dict(lang) for n in pagination.items],
        "total": pagination.total,
        "pages": pagination.pages,
        "current_page": page,
    }), 200


@news_bp.route("/<int:news_id>", methods=["GET"])
def get_news_item(news_id):
    lang = request.args.get("lang", "en")
    news = News.query.get_or_404(news_id)
    return jsonify(news.to_dict(lang)), 200


@news_bp.route("/admin", methods=["GET"])
@jwt_required()
def get_all_news_admin():
    """Get all news for admin (including unpublished)."""
    page = request.args.get("page", 1, type=int)
    per_page = request.args.get("per_page", 10, type=int)

    query = News.query.order_by(News.created_at.desc())
    pagination = query.paginate(page=page, per_page=per_page, error_out=False)

    return jsonify({
        "news": [n.to_dict("en") for n in pagination.items],
        "total": pagination.total,
        "pages": pagination.pages,
        "current_page": page,
    }), 200


@news_bp.route("", methods=["POST"])
@jwt_required()
def create_news():
    data = request.get_json()
    if not data:
        return jsonify({"error": "No data provided"}), 400

    news = News(
        title_en=data.get("title_en"),
        title_ar=data.get("title_ar"),
        content_en=data.get("content_en"),
        content_ar=data.get("content_ar"),
        region=data.get("region", "both"),
        image_url=data.get("image_url"),
        published=data.get("published", False),
        published_at=datetime.utcnow() if data.get("published") else None,
        player_id=data.get("player_id"),
    )
    db.session.add(news)
    db.session.commit()
    return jsonify(news.to_dict()), 201


@news_bp.route("/<int:news_id>", methods=["PUT"])
@jwt_required()
def update_news(news_id):
    news = News.query.get_or_404(news_id)
    data = request.get_json()

    for field in ["title_en", "title_ar", "content_en", "content_ar", "region", "image_url", "player_id"]:
        if field in data:
            setattr(news, field, data[field])

    if "published" in data:
        was_published = news.published
        news.published = data["published"]
        if data["published"] and not was_published:
            news.published_at = datetime.utcnow()

    db.session.commit()
    return jsonify(news.to_dict()), 200


@news_bp.route("/<int:news_id>", methods=["DELETE"])
@jwt_required()
def delete_news(news_id):
    news = News.query.get_or_404(news_id)
    db.session.delete(news)
    db.session.commit()
    return jsonify({"message": "News deleted"}), 200

File: server/src/routes/news.ts
import { Router } from 'express';
import { z } from 'zod';
import { NEWS_REGIONS, News } from '../models/index.js';
import { asyncHandler } from '../lib/async-handler.js';
import { HttpError } from '../lib/http-error.js';
import { objectId, pagination, parseBody, parseQuery } from '../lib/validate.js';
import { normalizeLang, paginationMeta, serializeNews } from '../lib/serializers.js';
import { requireAdmin } from '../middleware/auth.js';
import { writeLimiter } from '../middleware/rate-limit.js';

export const newsRouter: Router = Router();

const newsBody = z.object({
  title_en: z.string().max(500).nullish(),
  title_ar: z.string().max(500).nullish(),
  content_en: z.string().max(100_000).nullish(),
  content_ar: z.string().max(100_000).nullish(),
  region: z.enum(NEWS_REGIONS).optional(),
  image_url: z.string().max(500).nullish(),
  published: z.boolean().optional(),
  is_featured: z.boolean().optional(),
  player_id: z.union([objectId, z.literal(''), z.null()]).optional(),
});

type NewsBody = z.infer<typeof newsBody>;

function toAttrs(body: Partial<NewsBody>) {
  const attrs: Record<string, unknown> = {};
  if (body.title_en !== undefined) attrs.titleEn = body.title_en || null;
  if (body.title_ar !== undefined) attrs.titleAr = body.title_ar || null;
  if (body.content_en !== undefined) attrs.contentEn = body.content_en || null;
  if (body.content_ar !== undefined) attrs.contentAr = body.content_ar || null;
  if (body.region !== undefined) attrs.region = body.region;
  if (body.image_url !== undefined) attrs.imageUrl = body.image_url || null;
  if (body.player_id !== undefined) attrs.playerId = body.player_id || null;
  return attrs;
}

/** Only one article may be featured at a time. */
async function makeExclusiveFeature(keepId: string) {
  await News.updateMany({ _id: { $ne: keepId }, isFeatured: true }, { $set: { isFeatured: false } });
}

// ------------------------------------------------------------------ public

const listQuery = pagination(10).extend({
  lang: z.enum(['en', 'ar']).default('en'),
  player_id: objectId.optional(),
});

newsRouter.get(
  '/',
  asyncHandler(async (req, res) => {
    const { page, per_page: perPage, lang, player_id: playerId } = parseQuery(listQuery, req);

    // An article is visible in a language edition if it targets that language
    // or both.
    const filter: Record<string, unknown> = {
      published: true,
      region: { $in: [lang, 'both'] },
    };
    if (playerId) filter.playerId = playerId;

    const [articles, total] = await Promise.all([
      News.find(filter)
        .sort({ publishedAt: -1, createdAt: -1 })
        .skip((page - 1) * perPage)
        .limit(perPage)
        // Populating here is what keeps the list to two queries instead of one
        // per article for the author's name.
        .populate('playerId', 'nameEn nameAr title'),
      News.countDocuments(filter),
    ]);

    res.json({
      news: articles.map((article) => serializeNews(article, lang)),
      ...paginationMeta({ page, perPage, total }),
    });
  }),
);

// Must be declared before "/:id" or "admin" would be parsed as an id.
newsRouter.get(
  '/admin',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { page, per_page: perPage } = parseQuery(pagination(10), req);

    const [articles, total] = await Promise.all([
      News.find()
        .sort({ createdAt: -1 })
        .skip((page - 1) * perPage)
        .limit(perPage)
        .populate('playerId', 'nameEn nameAr title'),
      News.countDocuments(),
    ]);

    res.json({
      news: articles.map((article) => serializeNews(article, 'en')),
      ...paginationMeta({ page, perPage, total }),
    });
  }),
);

newsRouter.get(
  '/:id',
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const lang = normalizeLang(req.query.lang);

    const article = await News.findById(id).populate('playerId', 'nameEn nameAr title');
    if (!article) throw HttpError.notFound('Article not found');

    // Drafts are visible to admins only; to everyone else they do not exist.
    if (!article.published && req.auth?.role !== 'admin') {
      throw HttpError.notFound('Article not found');
    }

    res.json(serializeNews(article, lang));
  }),
);

// ------------------------------------------------------------------- admin

newsRouter.post(
  '/',
  requireAdmin,
  writeLimiter,
  asyncHandler(async (req, res) => {
    const body = parseBody(newsBody, req);
    const published = body.published ?? false;

    const article = await News.create({
      ...toAttrs(body),
      published,
      isFeatured: body.is_featured ?? false,
      publishedAt: published ? new Date() : null,
    });

    if (article.isFeatured) await makeExclusiveFeature(String(article._id));
    await article.populate('playerId', 'nameEn nameAr title');

    res.status(201).json(serializeNews(article, 'en'));
  }),
);

newsRouter.put(
  '/:id',
  requireAdmin,
  writeLimiter,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const body = parseBody(newsBody.partial(), req);

    const article = await News.findById(id);
    if (!article) throw HttpError.notFound('Article not found');

    article.set(toAttrs(body));

    if (body.published !== undefined) {
      // Stamp publishedAt the first time it goes live, and keep that original
      // date across later edits.
      if (body.published && !article.published) article.publishedAt = new Date();
      article.published = body.published;
    }
    if (body.is_featured !== undefined) article.isFeatured = body.is_featured;

    await article.save();
    if (article.isFeatured) await makeExclusiveFeature(id);
    await article.populate('playerId', 'nameEn nameAr title');

    res.json(serializeNews(article, 'en'));
  }),
);

newsRouter.delete(
  '/:id',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const deleted = await News.findByIdAndDelete(id);
    if (!deleted) throw HttpError.notFound('Article not found');

    res.json({ message: 'Article deleted' });
  }),
);
